### PR TITLE
PCH Settings: Refactor the client-side Settings API

### PR DESCRIPTION
**Pull request by vaurdan**
## Description
This update involves some refactoring to the settings management within the content helper components. We have removed the `useSaveSettings` hook and its associated functionality from the hooks directory. Instead, we introduced a more centralized settings management system, which includes a new settings provider (`SettingsProvider`), a custom hook (`useSettings`), and a dedicated settings store. This new system is designed to improve the manageability and scalability of settings across different components, ensuring a more consistent and efficient handling of user preferences and configurations.

### How to use this new Settings API?

To use the new settings API in a new component, follow these steps:

### 1. Wrap your component with `SettingsProvider`

This provider supplies the settings context to all child components. You'll need to specify the `endpoint` (a unique identifier for your settings) and `defaultSettings` (the initial settings values).

```jsx
<SettingsProvider endpoint="your-endpoint" defaultSettings={defaultSettings}>
  <YourComponent />
</SettingsProvider>
```

#### Using `defaultSettings` with `getSettingsFromJson`

The `defaultSettings` parameter in `SettingsProvider` can be dynamically generated using a custom function like `getSettingsFromJson`. This function can parse a JSON string (potentially coming from a server-side or initial configuration) and convert it into a JavaScript object that matches the structure expected by your component's settings.

Here's a brief example:

```jsx
// A custom function to parse settings JSON and provide defaults
const getSettingsFromJson = (settingsJson) => {
  try {
    return JSON.parse(settingsJson);
  } catch (error) {
    // Return default settings in case of an error
    return {
      setting1: 'default1',
      setting2: 'default2',
    };
  }
};

// Usage with SettingsProvider
<SettingsProvider endpoint="your-endpoint" defaultSettings={getSettingsFromJson(yourSettingsJson)}>
  <YourComponent />
</SettingsProvider>
```

This approach allows for flexibility in initializing your component's settings, especially when dealing with server-provided configurations or needing to ensure sensible defaults.

### 2. Use `useSettings` within your component

This custom hook allows you to access and modify the settings within any component nested under `SettingsProvider`.

```jsx
import { useSettings } from '../common/settings';

const YourComponent = () => {
  const { settings, setSettings } = useSettings();

  // Use `settings` to access current settings
  // Use `setSettings` to update settings, e.g., setSettings({ newSetting: value });

  return (
    // Your component JSX
  );
};
```

### 3. Saving settings

Changes made via `setSettings` are automatically persisted based on the `endpoint` you specified in the `SettingsProvider`. There's no need to manually save settings or call an API; the context provider and hook manage this for you.

### 4. Adding a New Setting Type

When introducing a new type of setting for your component, you should define its structure as an interface in the `settings/types` directory. This ensures type safety and clarity across the application. Follow these steps to add a new setting type:

1. **Create a new type file**: In the `settings/types` directory, create a new TypeScript file for your settings type. The file name should reflect the settings type, e.g., `my-new-settings.d.ts`.

2. **Define the interface**: Within this file, define an interface that represents the structure of your new settings. For example:

```typescript
export interface MyNewSettings {
  setting1: string;
  setting2: boolean;
  setting3: number;
}
```

3. **Export the interface**: Make sure to export this interface from the file so it can be imported elsewhere in your application.

4. **Update the main types file**: In the `settings/types/index.ts` file, import your new interface and then export it again. This makes it accessible via a single import from the `settings/types` directory.

```typescript
import type { MyNewSettings } from './my-new-settings';

export type {
  MyNewSettings,
  // other settings types...
};

// Don't forget to add your new type to the Settings type union
export type Settings = MyNewSettings | SidebarSettings | TopPostsSettings;
```

5. **Use the new settings type**: Now you can import your new settings type wherever needed in your components or other parts of the application to ensure type safety and clarity when working with these settings.

```jsx
import { MyNewSettings } from '../settings';

const YourComponent = () => {
  const { settings, setSettings } = useSettings<MyNewSettings>();

  // Now `settings` and the argument to `setSettings` are typed as `MyNewSettings`
};
```

By following these steps, you can systematically introduce new settings types, maintaining a clear and maintainable structure for managing component configurations.

## Motivation and context
The motivation behind these changes was to address some limitations of the previous settings management approach, which relied on individual components to manage and persist their settings. This was leading to redundant code, potential inconsistencies, and challenges in maintaining and extending settings functionality. By centralizing settings management, we aim to simplify the codebase, improve reusability, and provide a clear and consistent API for managing settings across all components.

## How has this been tested?
The changes have been tested across all affected components to ensure that settings are correctly initialized, managed, and persisted. I have done some manual testing within the application to verify that user settings are correctly applied and saved across sessions.

## Screenshots (if appropriate)
N/A

### CODE REVIEW

File: src/content-helper/common/settings/provider.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+	deps: ReactDeps = undefined
+) => {
+	const isFirstRender = useRef( true );
+
+	useEffect( () => {
+		// Don't save settings on the first render.
+		if ( isFirstRender.current ) {
+			isFirstRender.current = false;
+			return;
+		}
```
Comment (acicovic) [inline src/content-helper/common/settings/provider.tsx:98]: Just wanted to verify whether this is still needed with the new implementation.

File: src/content-helper/common/settings/store.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+// Union of all action types.
+type ActionTypes = SetSettingsAction | SetPartialSettingsAction;
+
+// Default state.
+const DEFAULT_STATE: SettingsState = {};
+
+// Store name.
+export const STORE_NAME = 'wpparsely/settings';
```
Comment (acicovic) [inline src/content-helper/common/settings/store.ts]: Wondering whether we should put a hyphen (`wp-parsely`) here or not. No strong opinion, just checking.

**Review by acicovic — COMMENTED** [review]
Great to see this re-worked and taken to the next level!

Left a couple of questions for now, and I'll also test that manually a bit.

File: src/content-helper/common/settings/store.ts (hunk 2 above)
Comment (vaurdan) [reply src/content-helper/common/settings/store.ts]: That's a good call, the other stores we use have a hyphen, so for consistency sake makes perfect sense to use `wp-parsely/settings` here.

File: src/content-helper/common/settings/provider.tsx (hunk 1 above)
Comment (vaurdan) [reply src/content-helper/common/settings/provider.tsx:98]: I verified it now, and it still needs this check.

**Pull request merged**